pcdscalc/diffraction.py

```diff
@@ -118,3 +121,109 @@ def d_space(material_id, hkl):
     )
     d = inv_d_sqr ** -0.5
     return d
+
+
+def form_factor(material_id, twotheta, energy):
+    """
+    Returns the atomic form factor for Rayleigh scattering
+
+    Parameters
+    ----------
+    material_id: str
+    twotheta: float
+        scattering angle in degrees
+    energy: float
+    """
+    material_id = cst.chemical_name_to_formula.get(material_id, material_id)
+    z = cst.element_z[material_id]
+    q = xraylib.MomentTransf(energy*1e-3, np.deg2rad(twotheta/2))
+    f = xraylib.FF_Rayl(z, q)
+    return f
+
+
+def structure_factor(material_id, f, hkl, z=None):
+    material_id = cst.chemical_name_to_formula.get(material_id, material_id)
+    lattice = cst.lattice_type[material_id]
+    h, k, l = hkl  # noqa: E741
+
+    if lattice == 'fcc':
+        x1 = np.exp(-1j*np.pi * (k+l))
+        x2 = np.exp(-1j*np.pi * (h+l))
+        x3 = np.exp(-1j*np.pi * (h+k))
+        F = f * (1 + x1 + x2 + x3)
+    elif lattice == 'bcc':
+        F = f * (1 + np.exp(-1j*np.pi * (h+k+l)))
+    elif lattice == 'cubic':
+        F = f
+    elif lattice == 'diamond':
+        x1 = np.exp(-1j*np.pi * (k+l))
+        x2 = np.exp(-1j*np.pi * (h+l))
+        x3 = np.exp(-1j*np.pi * (h+k))
+        x4 = np.exp(-1j*2*np.pi * (h/4.0 + k/4.0 + l/4.0))
+        F = f * (1 + x1 + x2 + x3) * (1 + x4)
+    elif lattice == 'rhomb':
+        z = cst.latticeParamRhomb[material_id]
+        F = f * (1 + np.exp(2*1j*np.pi * (h+k+l) * z))
+    elif lattice == 'tetr':
+        F = f
+    elif lattice == 'hcp':
+        F = f * (1 + np.exp(2*1j*np.pi * (h/3.0 + 2*k/3.0 + l/2.0)))
+    return F
+
+
+def unit_cell_volume(material_id):
+    """
+    Unit cell volume
+    """
+    material_id = cst.chemical_name_to_formula.get(material_id, material_id)
+    lp = cst.lattice_parameters[material_id]
+    # a, b, c in angstroms
+    # alpha, beta, gamma in degrees
+    a, b, c, alpha, beta, gamma = lp
+    a = a / units["ang"]
+    b = b / units["ang"]
+    c = c / units["ang"]
+    cos_alpha = cosd(alpha)
+    cos_beta = cosd(beta)
+    cos_gamma = cosd(gamma)
+    r = 1 - cos_alpha**2 - cos_beta**2 - cos_gamma**2 \
+        + 2 * cos_alpha * cos_beta * cos_gamma
+    volume = a * b * c * np.sqrt(r)
+    return volume
+
+
+def darwin_width(material_id, hkl, energy, T=293):
```

Comment:
Would it be possible/easy to add a few example inputs/outputs from the old python to the tests module as test cases here?
https://github.com/pcdshub/pcdscalc/blob/master/pcdscalc/tests/test_diffraction.py

Reply:
Done